File: algorithms/cracking_the_coding_interview/dfs_connected_cell_in_a_grid.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func solve(grid [][]int) int {
	max := 0

	for i, row := range grid {
		for j, value := range row {
			if value == 1 {
				count := countRegion(grid, i, j)
				if count > max {
					max = count
				}
			}
		}
	}

	return max
}

func countRegion(grid [][]int, i, j int) int {
	grid[i][j] = 0
	count := 1

	gridRows := len(grid)
	gridColumns := len(grid[0])

	for _, ii := range []int{-1, 0, 1} {
		for _, jj := range []int{-1, 0, 1} {
			// if the same element
			if ii == 0 && jj == 0 {
				continue
			}

			newI := i + ii
			newJ := j + jj

			if newI < 0 || newI >= gridRows {
				continue
			}

			if newJ < 0 || newJ >= gridColumns {
				continue
			}

			if value := grid[newI][newJ]; value == 1 {
				count += countRegion(grid, newI, newJ)
			}
		}
	}

	return count
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	numberOfRowsLine, _ := reader.ReadString('\n')
	numberOfRows, _ := strconv.Atoi(strings.TrimSpace(numberOfRowsLine))
	reader.ReadString('\n')

	grid := make([][]int, numberOfRows)
	for i := 0; i < numberOfRows; i++ {
		rowLine, _ := reader.ReadString('\n')
		rowLine = strings.TrimSpace(rowLine)
		rowAsString := strings.Fields(rowLine)
		row := make([]int, len(rowAsString))
		for j := range rowAsString {
			row[j], _ = strconv.Atoi(rowAsString[j])
		}
		grid[i] = row
	}
	fmt.Println(solve(grid))
}

File: algorithms/cracking_the_coding_interview/dfs_connected_cell_in_a_grid.php
<?php
declare(strict_types=1);

/**
 * @param int[][] $grid
 * @return int
 */
function solve(array $grid): int
{
    $max = 0;

    foreach ($grid as $i => $row) {
        foreach ($row as $j => $value) {
            if ($value === 1) {
                $count = countRegion($grid, $i, $j);
                if ($count > $max) {
                    $max = $count;
                }
            }
        }
    }

    return $max;
}

/**
 * @param int[][] $grid
 * @param int $i
 * @param int $j
 * @return int
 */
function countRegion(array &$grid, int $i, int $j): int
{
    $grid[$i][$j] = 0;
    $count = 1;

    foreach ([-1, 0, 1] as $ii) {
        foreach ([-1, 0, 1] as $jj) {
            // if the same element
            if ($ii === 0 && $jj === 0) {
                continue;
            }

            $newI = $i + $ii;
            $newJ = $j + $jj;

            if (isset($grid[$newI][$newJ]) && $grid[$newI][$newJ] === 1) {
                $count += countRegion($grid, $newI, $newJ);
            }
        }
    }

    return $count;
}

$handle = fopen('php://stdin', 'r');
fscanf($handle, '%d', $n);
fscanf($handle, '%d', $m);

$grid = [];
for($i = 0; $i < $n; $i++) {
    $grid[$i] = explode(' ', fgets($handle));
    $grid[$i] = array_map(function($value) {
        return (int) $value;
    }, $grid[$i]);
}
echo solve($grid);
